File: csharp/BronKerbosch/BronKerbosch3GP.cs
// Bron-Kerbosch algorithm with degeneracy ordering,
// with nested searches choosing a pivot from candidates only (IK_GP).

using BronKerbosch;
using System.Collections.Generic;

internal static class BronKerbosch3GP<VertexSet, VertexSetMgr>
    where VertexSet : ISet<Vertex>
    where VertexSetMgr : IVertexSetMgr<VertexSet>
{
#   pragma warning disable IDE0022 // Use expression body for method
    public static void Explore(UndirectedGraph<VertexSet, VertexSetMgr> graph, ICliqueConsumer consumer)
    {
        BronKerboschDegeneracy<VertexSet, VertexSetMgr>.Explore(graph, consumer, PivotChoice.MaxDegreeLocal);
    }
}

File: fsharp/BronKerbosch/BronKerbosch3GP.fs
// Bron-Kerbosch algorithm with degeneracy ordering,
// with nested searches choosing a pivot from candidates only (IK_GP).
module BronKerbosch3GP

open BronKerbosch

let public explore (graph: UndirectedGraph) (consumer: CliqueConsumer) : Unit =
    DegeneracyBased.explore MaxDegreeLocal graph consumer

let algorithm: Algorithm = { name = "Ver3½-GP"; exec = explore }
